presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -831,9 +880,92 @@ public long getTimeSpentWaitingForOutputInMillis()
         }
     }
 
+    private static class DiskPageOutputSupplier
+            implements OutputSupplier<PrestoSparkDiskPage>
+    {
+        private static final MutablePartitionId DEFAULT_PARTITION = new MutablePartitionId();
+
+        private final PrestoSparkOutputBuffer<PrestoSparkBufferedSerializedPage> outputBuffer;
+        private final TempStorage tempStorage;
+        private final TempDataOperationContext tempDataOperationContext;
+        private TempDataSink tempDataSink;
+
+        private long timeSpentWaitingForOutputInMillis;
+
+        private DiskPageOutputSupplier(PrestoSparkOutputBuffer<PrestoSparkBufferedSerializedPage> outputBuffer,
+                TempStorage tempStorage,
+                TempDataOperationContext tempDataOperationContext)
+        {
+            this.outputBuffer = requireNonNull(outputBuffer, "outputBuffer is null");
+            this.tempStorage = requireNonNull(tempStorage, "tempStorage is null");
+            this.tempDataOperationContext = requireNonNull(tempDataOperationContext, "tempDataOperationContext is null");
+
+            try {
+                tempStorage.initialize(false);
+            }
+            catch (IOException e) {
+                throw new RuntimeException("Unable to initialize tempStorage: ", e);
+            }
+        }
+
+        @Override
+        public Tuple2<MutablePartitionId, PrestoSparkDiskPage> getNext()
+                throws InterruptedException
+        {
+            long start = System.currentTimeMillis();
+            PrestoSparkBufferedSerializedPage page = outputBuffer.get();
+            if (page == null) {
+                return null;
+            }
+
+            int compressedBroadcastSizeInBytes = 0;
+            int uncompressedBroadcastSizeInBytes = 0;
+            int positionCount = 0;
+            List<Long> pageChecksum = new ArrayList<>();
+            TempStorageHandle tempStorageHandle;
+
+            try {
+                this.tempDataSink = tempStorage.create(tempDataOperationContext);
+                while (page != null) {
+                    tempDataSink.write(ImmutableList.of(new PageDataOutput(page.getSerializedPage())));
```

Comment:
Sometimes pages could be very small. From what I remember the storage implementation is not required to buffer. If that's the case  - we should buffer up to some amount here. For PageFile format we buffer up to 24mb by default (that's an optimal block size for tempfs, but it is configurable).

CC: @wenleix @viczhang861